In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

from sklearn.model_selection import train_test_split

from sklearn.linear_model import (
    LinearRegression,
    Ridge,
    Lasso,
    ElasticNet
)

from sklearn.tree import DecisionTreeRegressor
from sklearn.ensemble import RandomForestRegressor
from sklearn.svm import SVR

from sklearn.metrics import (
    mean_absolute_error,
    mean_squared_error,
    r2_score
)

In [3]:
df = pd.read_csv("processed_telco.csv")

In [4]:
X = df.drop("MonthlyCharges", axis=1)
y = df["MonthlyCharges"]

In [5]:
X_train, X_test, y_train, y_test = train_test_split(
    X,
    y,
    test_size=0.2,
    random_state=42
)

In [6]:
lr = LinearRegression()

lr.fit(X_train, y_train)

y_pred_lr = lr.predict(X_test)

In [7]:
print("MAE:", mean_absolute_error(y_test, y_pred_lr))
print("MSE:", mean_squared_error(y_test, y_pred_lr))
print("RMSE:", np.sqrt(mean_squared_error(y_test, y_pred_lr)))
print("R2 Score:", r2_score(y_test, y_pred_lr))

MAE: 0.7886182909621687
MSE: 1.1008749287814594
RMSE: 1.0492258711933573
R2 Score: 0.9987840669478372


Linear Regression achieved an R² score of 0.9988, indicating that it explained nearly all variability in MonthlyCharges. The model also achieved very low MAE, MSE, and RMSE values, demonstrating highly accurate predictions.

In [8]:
ridge = Ridge(alpha=1.0)

ridge.fit(X_train, y_train)

y_pred_ridge = ridge.predict(X_test)

In [9]:
print("MAE:", mean_absolute_error(y_test, y_pred_ridge))
print("MSE:", mean_squared_error(y_test, y_pred_ridge))
print("RMSE:", np.sqrt(mean_squared_error(y_test, y_pred_ridge)))
print("R2 Score:", r2_score(y_test, y_pred_ridge))

MAE: 0.7890679202860219
MSE: 1.1009982802522151
RMSE: 1.0492846516804748
R2 Score: 0.9987839307042673


Ridge Regression produced nearly identical results to Linear Regression.
This indicates that the dataset does not suffer significantly from overfitting or multicollinearity, and both models fit the data extremely well.

Ridge Regression achieved performance nearly identical to Linear Regression, with an R² score of 0.9988. The regularization term did not significantly improve performance, suggesting that the original Linear Regression model was already highly stable.

In [10]:
from sklearn.linear_model import Lasso
lasso = Lasso(alpha=0.01)
lasso.fit(X_train, y_train)
y_pred_lasso = lasso.predict(X_test)

In [11]:
print("MAE:", mean_absolute_error(y_test, y_pred_lasso))
print("MSE:", mean_squared_error(y_test, y_pred_lasso))
print("RMSE:", np.sqrt(mean_squared_error(y_test, y_pred_lasso)))
print("R2 Score:", r2_score(y_test, y_pred_lasso))

MAE: 0.7915227713164436
MSE: 1.1017512184659293
RMSE: 1.0496433768027735
R2 Score: 0.9987830990725929


Lasso Regression achieved an R² score of 0.9988, which is comparable to Linear and Ridge Regression. The slightly higher MAE and RMSE indicate a marginal reduction in predictive performance due to coefficient regularization.

In [12]:
from sklearn.linear_model import ElasticNet

elastic = ElasticNet(
    alpha=0.01,
    l1_ratio=0.5,
    random_state=42
)

elastic.fit(X_train, y_train)

y_pred_elastic = elastic.predict(X_test)

In [13]:
print("MAE:", mean_absolute_error(y_test, y_pred_elastic))
print("MSE:", mean_squared_error(y_test, y_pred_elastic))
print("RMSE:", np.sqrt(mean_squared_error(y_test, y_pred_elastic)))
print("R2 Score:", r2_score(y_test, y_pred_elastic))

MAE: 0.8679338221312098
MSE: 1.2556958059882473
RMSE: 1.1205783354983476
R2 Score: 0.9986130649413069


Linear Regression remains the best-performing linear model.
ElasticNet has the highest error among the linear models.

In [14]:
from sklearn.tree import DecisionTreeRegressor

dt_reg = DecisionTreeRegressor(
    random_state=42
)

dt_reg.fit(X_train, y_train)

y_pred_dt = dt_reg.predict(X_test)

In [15]:
print("MAE:", mean_absolute_error(y_test, y_pred_dt))
print("MSE:", mean_squared_error(y_test, y_pred_dt))
print("RMSE:", np.sqrt(mean_squared_error(y_test, y_pred_dt)))
print("R2 Score:", r2_score(y_test, y_pred_dt))

MAE: 1.3291341376863024
MSE: 4.251889638041164
RMSE: 2.062011066420635
R2 Score: 0.9953037234204567


Decision Tree Regressor performed worse than all linear models.
It produced higher MAE and RMSE values, indicating larger prediction errors.

In [16]:
from sklearn.ensemble import RandomForestRegressor

rf_reg = RandomForestRegressor(
    n_estimators=100,
    random_state=42
)

rf_reg.fit(X_train, y_train)

y_pred_rf = rf_reg.predict(X_test)

In [17]:
print("MAE:", mean_absolute_error(y_test, y_pred_rf))
print("MSE:", mean_squared_error(y_test, y_pred_rf))
print("RMSE:", np.sqrt(mean_squared_error(y_test, y_pred_rf)))
print("R2 Score:", r2_score(y_test, y_pred_rf))

MAE: 0.9761628814762225
MSE: 2.1958558816536526
RMSE: 1.4818420569189055
R2 Score: 0.9975746438814406


Random Forest Regressor improved significantly over Decision Tree Regressor but still did not outperform the linear models. Linear Regression continues to achieve the lowest prediction error and highest R² score.

In [18]:
from sklearn.svm import SVR

In [19]:
svr = SVR(
    kernel='rbf'
)

svr.fit(X_train, y_train)

y_pred_svr = svr.predict(X_test)

In [20]:
print("MAE:", mean_absolute_error(y_test, y_pred_svr))
print("MSE:", mean_squared_error(y_test, y_pred_svr))
print("RMSE:", np.sqrt(mean_squared_error(y_test, y_pred_svr)))
print("R2 Score:", r2_score(y_test, y_pred_svr))

MAE: 19.074176008011676
MSE: 519.8914486035223
RMSE: 22.80112823093459
R2 Score: 0.42577201154581357


Comparison Table

In [21]:
comparison_reg = pd.DataFrame({
    'Model': [
        'Linear Regression',
        'Ridge Regression',
        'Lasso Regression',
        'ElasticNet',
        'Decision Tree Regressor',
        'Random Forest Regressor',
        'SVR'
    ],
    'MAE': [
        0.7886,
        0.7891,
        0.7915,
        0.8680,
        1.3291,
        0.9762,
        19.0742
    ],
    'RMSE': [
        1.0492,
        1.0493,
        1.0497,
        1.1206,
        2.0620,
        1.4818,
        22.8011
    ],
    'R2 Score': [
        0.99878,
        0.99878,
        0.99878,
        0.99861,
        0.99530,
        0.99757,
        0.42577
    ]
})

comparison_reg

,Model,MAE,RMSE,R2 Score
0,Linear Regression,0.7886,1.0492,0.99878
1,Ridge Regression,0.7891,1.0493,0.99878
2,Lasso Regression,0.7915,1.0497,0.99878
3,ElasticNet,0.8680,1.1206,0.99861
4,Decision Tree Regressor,1.3291,2.0620,0.99530
5,Random Forest Regressor,0.9762,1.4818,0.99757
6,SVR,19.0742,22.8011,0.42577


Among all regression models, Linear Regression achieved the best overall performance with the lowest MAE (0.7886), lowest RMSE (1.0492), and highest R² score (0.9988). Ridge and Lasso Regression produced nearly identical results, while Decision Tree and Random Forest Regressors showed slightly higher prediction errors. SVR performed significantly worse than the other models and achieved the lowest R² score.